Core/UsageSegmentation.swift

```diff
@@ -0,0 +1,100 @@
+//
+//  UsageSegmentation.swift
+//  DuckDuckGo
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+enum UsageActivityType: String {
+
+    case search
+    case appUse = "app_use"
+
+}
+
+protocol UsageSegmenting {
+
+    func processATB(_ atb: Atb, withInstallAtb installAtb: Atb, andActivityType activityType: UsageActivityType)
+
+}
+
+class UsageSegmentation: UsageSegmenting {
+
+    private let pixelFiring: PixelFiring.Type
+    private let storage: UsageSegmentationStoring
+    private let calculatorFactory: UsageSegmentationCalculatorMaking
+
+    init(pixelFiring: PixelFiring.Type = Pixel.self,
+         storage: UsageSegmentationStoring = UsageSegmentationStorage(),
+         calculatorFactory: UsageSegmentationCalculatorMaking = DefaultCalculatorFactory()) {
+        self.pixelFiring = pixelFiring
+        self.storage = storage
+        self.calculatorFactory = calculatorFactory
+    }
+
+    func processATB(_ atb: Atb, withInstallAtb installAtb: Atb, andActivityType activityType: UsageActivityType) {
+        var atbs = activityType.atbsFromStorage(storage)
+
+        // Fail fast by looking at the end of the list
+        guard !atbs.reversed().contains(where: { $0 == atb }) else { return }
+
+        defer {
+            activityType.updateStorage(storage, withAtbs: atbs)
+        }
+
+        if atbs.isEmpty {
+           atbs.append(installAtb)
+        }
+
+        if installAtb != atb {
+           atbs.append(atb)
+        }
+
+        var pixelInfo: [String: String]?
+        let calculator = calculatorFactory.make(installAtb: installAtb)
+        
+        for atb in atbs {
+            pixelInfo = calculator.processAtb(atb, forActivityType: activityType)
```

Comment:
Is this intended to keep setting the `pixelInfo`? Is the loop required, or would this work:

```
if let lastAtb = atbs.last {
    pixelInfo = calculator.processAtb(lastAtb, forActivityType: activityType)
}
```

Reply:
Looking more at the calculator, I think this is intended, as the calculator updates it’s state on every call

Reply:
Yep, that's right.  I'll add a comment to make that clearer.